### PR TITLE
Swift: Better sinks for swift/cleartext-storage-database

### PR DESCRIPTION
**Pull request by geoffw0**
There are a couple of issues with the sensitive data queries that are preventing us from finding four results on DVIA-v2 such as [this one](https://github.com/prateek147/DVIA-v2/blob/master/DVIA-v2/DVIA-v2/Vulnerabilities/Insecure%20Data%20Storage/Controller/CoreDataViewController.swift#L44) , even though we fixed the `#available` extraction issue that was previously blocking these results.  This PR attempts to test and fix the newer issues:
- in cases like an access of `password.text`, the sensitive data heuristic identifies the container (`password`) as sensitive based on its name and not the access of the contents (`.text`).  The current solution for this is a (fairly wide) additional taint step.
- sometimes the sink models `NSManagedObject` and `RealmSwiftObject` don't work in a way I haven't been able to model in tests.  I think it may have something to do with the way the real world code for these types can be split between [different files](https://github.com/prateek147/DVIA-v2/blob/3aa63279969e60ea8180535cf7983795abfa4714/DVIA-v2/DVIA-v2/Model/CoreDataClasses/User+CoreDataProperties.swift#L13) that are compiled separately ???  The solution is to work through the type (`NominalType`) rather than the declaration (`ClassOrStructDecl`).

I'm not entirely happy with the overly general additional taint step.  I'm very much open to discussing your concerns and suggestions.  I have tried some alternatives:
 - using implicit reads for fields at the source
   - gets a few of the test results but not all of them
 - making both the qualifier and the member reference a sensitive expression, regardless of which triggered the sensitive data heuristic
   - again gets some cases but not others
 - both of the above
   - I was able to catch all test cases apart from the last one; didn't get the DVIA-v2 results either and I'm not sure why.
 - I briefly tried using `TaintInheritingContent` as well but was not able to get any of the test cases with that.

The DCA run should show four new results on DVIA-v2, and hopefully similar performance.

---

(after this is merged, I should extend both changes to similar queries; lets discuss what's here first)

### CODE REVIEW

**Comment by geoffw0** [comment]
DCA:
 - we get the 4 expected results on DVIA-v2 :tada:
 - we also get 1 new result on SideStore, which LGTM :tada:
 - we also get 3 new results on Signal-iOS, which are a bit less obvious...
    - in all cases the sensitive data is a field called `mobileCoin`, which the sensitive data library assumes to be a telephone number.  The truth appears to be that its an amount or type of currency, I think.  This seems like potentially sensitive data as well, but it's less clear cut than a phone number.
    - I'm not sure if these are actually values written to the database, or [keys looked up](https://github.com/signalapp/Signal-iOS/blob/2693c55232822d39fdbd03bde26cf7578e820f88/SignalServiceKit/src/Payments/PaymentFinder.swift#L142) in it.
      - this issue is specific to the GRDB sinks, because they always execute SQL so its difficult to tell what's really being done.
    - overall I think I would say these results are _probably_ false positives, though reasonable ones.
    - I will do a MRVA run and assess the quality of other GRDB results.
 - the result messages could probably be cleaner (by not targeting post-update node); I'll look into that as well.
 - nothing else stands out on DCA, in particular there are no indications of a performance issue.

Also some kind of mass autoformat error is going on in the CI (the autoformatter is undergoing an update).  I will merge main at some point and hope the issue goes away.

**Comment by groue** [comment]
>   * I'm not sure if these are actually values written to the database, or [keys looked up](https://github.com/signalapp/Signal-iOS/blob/2693c55232822d39fdbd03bde26cf7578e820f88/SignalServiceKit/src/Payments/PaymentFinder.swift#L142) in it.

This is a custom string interpolation defined by Signal: https://github.com/signalapp/Signal-iOS/blob/a81cb65e5b7906ae392249cac9d5c6fffe4096ba/SignalServiceKit/src/Payments/TSPaymentModel%2BSDS.swift#L109-L116

In this specific case, the custom interpolation injects a safe column name, not values.

But generally speaking, _anything is possible_. The code linked by "keys looked up" defines a Swift `String`, later interpreted as SQL: this string is fundamentally a vector for SQL injections:

- `"INSERT ... VALUES ('\(taintedInput)')"` (standard Swift)
- `"INSERT ... VALUES (\(myCustomInterpolation: taintedInput))"` (standard Swift with custom interpolation)

**Comment by geoffw0** [comment]
> This is a custom string interpolation defined by Signal

Thanks for this explanation, I hadn't understood exactly what was going on.

We currently have taint flow through all string interpolation, which seems like it's causing the issue here.  We should probably restrict that to default string interpolation.  It will still be possible to have flow through custom interpolations where appropriate, as CodeQL can see the Swift code for them and hopefully connect the dots ... or not, in cases like this.

**Comment by geoffw0** [comment]
I have fixed the result messages, merged with main, and all of the tests are passing now.

I've transitioned this PR out of draft.

I've also created two issues to follow up on different aspects of the `Signal-iOS` results.

File: swift/ql/lib/codeql/swift/security/CleartextStorageDatabaseExtensions.qll
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/security/CleartextStorageDatabaseExtensions.qll:140]: This basically introduces total field/object conflation for this query which may lead to a lot of FPs. Could we add it only at the sinks instead? Would that still catch the TPs we want to catch?

**Review by MathiasVP — APPROVED** [review]
Generally, LGTM. I'm a bit nervous about the generous new taint-step. If this is necessary for the new TPs then I'm okay with merging it.

File: swift/ql/lib/codeql/swift/security/CleartextStorageDatabaseExtensions.qll
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/CleartextStorageDatabaseExtensions.qll:140]: I'm also a bit nervous about this, but I don't think adding it at the sinks is enough (maybe at the sources???).  I'm fairly sure I _can_ replace `LookupExpr` with something a bit narrower, but that will still be 'total conflation' in a smaller class of things.  I'll experiment a bit more...

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/CleartextStorageDatabaseExtensions.qll:140]: (we will want this new taint step on the other sensitive data queries as well when we're done here, so it's worth a bit of time getting it right...)

**Comment by CasandraDevivaW1i** [comment]
It seems that there are a few issues with the sensitive data queries that are causing problems in finding results on DVIA-v2. The first issue is related to how the sensitive data heuristic identifies containers based on their names instead of their contents. The solution proposed is to use a wider taint step. The second issue is related to the way NSManagedObject and RealmSwiftObject work, which cannot be modeled in tests. The proposed solution is to work through the type rather than the declaration.

There have been some attempts to fix these issues, such as using implicit reads for fields at the source and making both the qualifier and the member reference a sensitive expression. However, these attempts have only caught some of the test cases and not all of them. The author is open to suggestions and discussions on how to improve these solutions.

In the end, the author hopes that the DCA run will show four new results on DVIA-v2 and similar performance.




Assistant03 Gua
Please reply to the conversation below comment
"There are a couple of issues with the sensitive data queries that are preventing us from finding four results on DVIA-v2 such as this one , even though we fixed the #available extraction issue that was previously blocking these results. This PR attempts to test and fix the newer issues:

in cases like an access of password.text, the sensitive data heuristic identifies the container (password) as sensitive based on its name and not the access of the contents (.text). The current solution for this is a (fairly wide) additional taint step.
sometimes the sink models NSManagedObject and RealmSwiftObject don't work in a way I haven't been able to model in tests. I think it may have something to do with the way the real world code for these types can be split between different files that are compiled separately ??? The solution is to work through the type (NominalType) rather than the declaration (ClassOrStructDecl).
I'm not entirely happy with the overly general additional taint step. I'm very much open to discussing your concerns and suggestions. I have tried some alternatives:

using implicit reads for fields at the source
gets a few of the test results but not all of them
making both the qualifier and the member reference a sensitive expression, regardless of which triggered the sensitive data heuristic
again gets some cases but not others
both of the above
I was able to catch all test cases apart from the last one; didn't get the DVIA-v2 results either and I'm not sure why.
I briefly tried using TaintInheritingContent as well but was not able to get any of the test cases with that.
The DCA run should show four new results on DVIA-v2, and hopefully similar performance."
It sounds like you've put a lot of effort into troubleshooting the issues with the sensitive data queries, and it's great to hear that you were able to catch most of the test cases. It's understandable that you're not entirely satisfied with the additional taint step, but it's good that you're open to discussing other options. Have you considered reaching out to your colleagues or team members for their input or suggestions? They may be able to offer some insights or ideas that could help improve the accuracy of the queries. Keep up the good work, and I hope the DCA run yields the desired results.

File: swift/ql/lib/codeql/swift/security/CleartextStorageDatabaseExtensions.qll
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/security/CleartextStorageDatabaseExtensions.qll]: I've narrowed the additional taint step from all `LookupExpr`s to just `MemberRefExpr`, and added a couple of additional test cases while I was thinking about this stuff.

I don't think a solution that works at the sinks can be good enough here.  I'm more interested in sources right now.  Consider:
```
var a = myPasswordBox
var b = a.textField
db.storedText = b
```
In this case the heuristic has determined `myPasswordBox` is sensitive, but *really* the sensitive data is `myPasswordBox.textField`.  What we really want is for all content of `myPasswordBox` on the first line to be considered tainted, then track where it goes.  But as far as I'm aware we don't have a tool for that (do we?), so we allow flow from `a` to `a.textField` instead.

What kinds of false positives worry you most?  I can imagine cases where a field of something we determine to be sensitive is actually not sensitive, but I think they will be unusual.  It's actually quite hard to think of examples, but I came up with:
- `password.length` should probably be sensitive IMO.
- `bankaccount.providerName` seems only a little bit sensitive, compared to `bankaccount.number` say.
- `mobilePhone.colour` wouldn't really be sensitive at all, though I think the real issue here is that matching `mobile` on its own is already dubious (I have an issue open to fix this).  Once that's done, the example would be `mobileNumber.colour` and that is nonsensical.  The idea here is that the sensitive data heuristic should be labelling fairly specific things, for which all fields would likely be sensitive.

Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/security/CleartextStorageDatabaseExtensions.qll]: I can see that having the additional step only at the sink is going to be limiting, indeed.

> But as far as I'm aware we don't have a tool for that (do we?),

Correct, we don't have such a tool. There's no way to insert such "implicit stores" similar to how we can insert "implicit reads".

> What kinds of false positives worry you most? I can imagine cases where a field of something we determine to be sensitive is actually not sensitive, but I think they will be unusual.

I haven't thought too carefully about it yet. I was thinking of strange cases like `mobilePhone.colour`, but when you say
> The idea here is that the sensitive data heuristic should be labelling fairly specific things, for which all fields would likely be sensitive.

I'm much less concerned about this additional step 👍.

**Review by MathiasVP — APPROVED** [review]
LGTM once DCA is happy!

**Pull request merged**
